### PR TITLE
DOM

### PR DESCRIPTION
**Pull request by kotlyar-andrey**
### DOM

[Demo](https://kotlyar-andrey.github.io/dom-ds/) |
[Code base](https://github.com/kotlyar-andrey/dom-ds)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

File: submissions/kotlyar-andrey/dom/script.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  const menu = document.querySelector(".menu");
+
+  menu.innerHTML = bosses
+    .map(
+      (boss) =>
+        `<a href="#" class="menu__item" id="menu_boss_${boss.id}">${boss.name}</a>`
+    )
+    .join("");
+  bosses.forEach((boss) => {
+    getMenuItem(boss.id).addEventListener("click", () => showBoss(boss.id));
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/script.js:112]: Use event delegation [read more](https://davidwalsh.name/event-delegate)

File: submissions/kotlyar-andrey/dom/script.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+      "Гвин был повелителем Света и отцом минимум четверых детей: первенца Солнца, Гвиневер, Гвиндолина и Филианоры. Йоршка, называющая себя сестрой Гвиндолина, также может быть ещё одной дочерью Гвина. Гвин и его дети были богами, однако первенец был позже лишен божественного статуса в наказание за заключение мира с Древними Драконами.",
+    video: "aM_OOcZVU9s",
+  },
+];
+
+/* Создание меню с добавлением прослушивания события нажатия */
+function createMenu() {
+  const menu = document.querySelector(".menu");
+
+  menu.innerHTML = bosses
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/script.js:105]: Use append method, which can add several elements in one operation

File: submissions/kotlyar-andrey/dom/script.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+  bosses.forEach((boss) => {
+    getMenuItem(boss.id).addEventListener("click", () => showBoss(boss.id));
+  });
+}
+
+/* Показ инфорации о выбранном боссе */
+function showBoss(id) {
+  const target = document.querySelector(".content");
+  const boss = bosses.filter((obj) => obj.id === id)[0];
+  target.innerHTML = `
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/script.js:120]: Use append method, which can add several elements in one operation

File: submissions/kotlyar-andrey/dom/script.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+/* Сохранение списка в куках */
+function setCookie() {
+  document.cookie = "completed=" + encodeURIComponent(completed);
+}
+
+/* Проверка, пройден ли босс. Если да, изменение цвета его отображения */
+function checkIsCompleted(id) {
+  if (completed.includes(id)) {
+    document.querySelector(".content__title").classList.add("completed");
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/script.js:184]: Mabey better to use `classList.toggle`?

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Awesome design!

File: submissions/kotlyar-andrey/dom/script.js (hunk 2 above)
Comment (kotlyar-andrey) [reply submissions/kotlyar-andrey/dom/script.js:105]:
```
  menu.innerHTML = bosses
    .map(
      (boss) =>
        `<a href="#" class="menu__item" id="menu_boss_${boss.id}" data-boss_id="${boss.id}">${boss.name}</a>`
    )
    .join("");
```
But I don't have several elements. I generate list of links, then join to string and then add  the html-string into menu. I don't have nodes for add them use append.

File: submissions/kotlyar-andrey/dom/script.js (hunk 3 above)
Comment (kotlyar-andrey) [reply submissions/kotlyar-andrey/dom/script.js:120]: Here is also one html block that I add to the container. I can't use append for this.

File: submissions/kotlyar-andrey/dom/script.js (hunk 2 above)
Comment (A-Ostrovnyy) [reply submissions/kotlyar-andrey/dom/script.js:105]: Each link is an element) also, `append` method is more productive, it increases page performance

**Comment by kotlyar-andrey** [comment]
Thanks for the hint with event delegation - it's really a good way.

I use `innerHtml` once for add **html code** into the target node. Method `append` is used for add **nodes** created with `document.createElement` or obtained with `document.getElement...`, so I really cant use it in my case. And I use  `innerHtml` once, not in the loop, so it won't have problems with performance.

File: submissions/kotlyar-andrey/dom/style.css [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+  padding: 0;
+}
+html,
+body {
+  height: 100%;
+  background-color: #212121;
+}
+@font-face {
+  font-family: darksouls;
+  src: url(fonts/CormorantUnicase-Regular.ttf);
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/style.css:12]: Avoid using `.ttf` fonts in the web use `woff2` along with `.woff` Use [this converter](https://transfonter.org/) for it

File: submissions/kotlyar-andrey/dom/script.js [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  }, 3000);
+  setTimeout(() => {
+    modal.remove();
+  }, 5000);
+}
+
+createMenu();
+
+const completed = getCookie();
+completed.forEach((id) => getMenuItem(id).classList.add("completed"));
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/script.js:238]: You can add class name when you create menu items

**Comment by kotlyar-andrey** [comment]
Thanks for the detailed review! I fixed everything. Re-review please)

File: submissions/kotlyar-andrey/dom/style.css [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+  margin: 0;
+  padding: 0;
+}
+html,
+body {
+  height: 100%;
+  background-color: #212121;
+}
+@font-face {
+  font-family: "Cormorant Unicase";
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/style.css:11]: Always add fallback font. [Read more](https://css-tricks.com/css-basics-fallback-font-stacks-robust-web-typography/)

File: submissions/kotlyar-andrey/dom/script.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  const menu = document.querySelector(".menu");
+
+  menu.innerHTML = bosses
+    .map(
+      (boss) =>
+        `<a href="#" class="menu__item${
+          completed.includes(boss.id) ? " completed" : ""
+        }" id="menu_boss_${boss.id}" data-bossid="${boss.id}">${boss.name}</a>`
+    )
+    .join("");
```
Comment (A-Ostrovnyy) [inline submissions/kotlyar-andrey/dom/script.js:112]: `[].map().join()` replace with `[].reduce((acc, item) => {}, '')` one method instead of two)

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Good progress!

**Comment by kotlyar-andrey** [comment]
Done!

File: submissions/kotlyar-andrey/dom/style.css (hunk 7 above)
Comment (A-Ostrovnyy) [reply submissions/kotlyar-andrey/dom/style.css:13]: Not fixed :cry:

**Review by A-Ostrovnyy — APPROVED** [review]
Well done!

**Pull request merged**